### DISCUSSION TITLE
NoUnreachableCodeFixer

### DISCUSSION THREAD

**Discussion by keradus** [PHP-CS-Fixer/PHP-CS-Fixer#7060] [RFC]
What do you think about fixer that would remove non-reachable code?

eg

before:

``` php
functinon foo()
{
    echo 1;
    return;
    echo 2;
}
```

after:

``` php
functinon foo()
{
    echo 1;
    return;
}
```

**Comment by hkdobrev** [comment]
👍  to have this. However I think this could be a risky fixer for some. It won't change what the code would do, but often unreachable code isn't intentional and people should actually run the code which is unreachable.

**Comment by keradus** [comment]
> often unreachable code isn't intentional

agree

> and people should actually run the code which is unreachable

why?

**Comment by phansys** [comment]
@keradus, I think it should be named "UnreachableCodeFixer" (instead of "NoUnreachableCodeFixer" which seems a contradiction for what you're proposing).

**Comment by keradus** [comment]
We name rules after state the code will follow after fixing. For all fixers. Just for consistency. So, after fixing, there should be no unreachable code.

**Comment by phansys** [comment]
Excellent then @keradus. Thanks for the clarification.

**Comment by SpacePossum** [comment]
``` php
             case 'directdebit':
                 return 'test';
                 break;
```

please let the fixer remove the `break` in these cases as well :)

**Comment by kubawerlos** [comment]
As much as having this fixer would be great I'm not sure it's worth the effort - would be very complicated and Vimeo's [Psalm](https://github.com/vimeo/psalm) has `--find-dead-code` option to do exactly this.

**Comment by VincentLanglet** [comment]
IMHO, it would be a risky fixer to automatically remove unreachable code. And I assume it wouldn't be easy to implement.

I would let static analysis doing this.
Psalm will do this in a better way because it supports the `never` return type.

**Comment by keradus** [comment]
> IMHO, it would be a risky fixer to automatically remove unreachable code.

can you share reasoning, please?

**Comment by keradus** [comment]
>  not sure it's worth the effort

> I assume it wouldn't be easy to implement.



I wouldn't focus myself on each edge case. But if we go with following: find `return` statement in non-nested body of function, if there is anything between it and end of function, drop it.

**Comment by VincentLanglet** [comment]
> > IMHO, it would be a risky fixer to automatically remove unreachable code.
> 
> can you share reasoning, please?

Sure, like it was said "often unreachable code isn't intentional"

I know multiple developers/teams running php-cs-fixer locally before committing every changes.

There is tools like phpstan which report an error when some code is unreachable.
Let's say you were debugging
```
public function foo(): void
{
     $this->doStart();
     die; // Was added for debugging purpose
     $this->doEnd();
}
```

If you run php-cs-fixer without this rule, then phpstan, you'll get an error and will catch your mistake.
If you run php-cs-fixer with this rule, then phpstan, you won't get any error since the code will be removed. It's risky.

**Comment by julienfalque** [comment]
One can disable the rule if it does not fit their workflow.

However, since static analysis tools can report dead code I don't think implementing this rule in PHP CS Fixer is worth the effort.

**Comment by keradus** [comment]
maybe debug code shouldn't be committed in first place. Yet as Julien mentioned, if one want's to do this, they may not enable the rule.

priority - fully agree, that's why nobody focused on it and implemented it. yet if one would want to - I wouldn't see harm in accepting it
